### PR TITLE
Rule S6419: Azure Functions should be stateless

### LINKED ISSUE #5595: Rule S6419: Azure Functions should be stateless
MMF: https://jira.sonarsource.com/browse/MMF-2699
RSPEC PR to be merged after implementing this ticket: https://github.com/SonarSource/rspec/pull/952

 * Azure Functions should not alter the state of any static class or the class that it's part of.

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #5595

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CloudNative/AzureFunctionsStateless.cs [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        --Field;    // Noncompliant
+        //^^^^^
+        Array[0]++; // Noncompliant
+        ++Array[0]; // Noncompliant
+
+        for (; Field < 100; Field++) { }    // Noncompliant
+    }
+
+    [FunctionName("Sample")]
+    public static void Assignment(object arg)
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CloudNative/AzureFunctionsStateless.cs:212]:
```suggestion
    public static void CompoundAssignment(object arg)
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CloudNative/AzureFunctionsStateless.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                    {
+                        CheckTarget(c, argument.Expression);
+                    }
+                },
+                SyntaxKind.Argument);
+        }
+
+        private static void CheckTarget(SyntaxNodeAnalysisContext context, ExpressionSyntax target)
+        {
+            if (context.IsAzureFunction()
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CloudNative/AzureFunctionsStateless.cs:88]: Could go into before the switch in the merged RegisterSyntaxNodeActionInNonGenerated calls.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CloudNative/AzureFunctionsStateless.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                && context.SemanticModel.GetSymbolInfo((target as ElementAccessExpressionSyntax)?.Expression ?? target).Symbol is { } symbol
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CloudNative/AzureFunctionsStateless.cs:89]: I would prefer to move the `(target as ElementAccessExpressionSyntax)?.Expression ?? target` target part to the caller. The caller already extracts the right expression from the node and should do so entirely if possible.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
I have some general concerns about the implementation.
1. The analyzer is called on a lot of nodes (almost every line of code contains one or more of the listed node kinds)
1. Mutation outside the azure function or in local functions is likely missed.

I would propose to register for method declarations and start a class wide analysis from there. This is triggered in just a fraction of cases and can bail out easily for all non az function declarations. This would find all mutations taking place inside the containing class (e.g. By searching for write operations to mutable static fields). The downside is, that we end up reporting diagnostics multiple times if there is more than one azure function per class. We could overcome this limitation by registering for class declarations instead and searching for FunctionName decorated methods from there.

For instance azure functions it should work by checking for mutation of static fields (I think, Azure function instances are created per request, so instance fields should be fine to mutate, but I haven't checked the life cycle of these things yet).

Further: While most samples use static azure functions, I think that this is not the way these functions are used these days in the majority of production code. DI and the options system are omnipresent in any nontrivial az function. See e.g. this article for why that is the case (I would say Key Vault, httpclient and IOptions are all very good reasons to use the asp DI system in az functions).
https://dev.to/shibayan/a-new-era-of-azure-functions-development-6f2

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CloudNative/AzureFunctionsStateless.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var argument = (ArgumentSyntax)c.Node;
+                    if (argument.RefOrOutKeyword != default)
+                    {
+                        CheckTarget(c, argument.Expression);
+                    }
+                },
+                SyntaxKind.Argument);
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CloudNative/AzureFunctionsStateless.cs:83]: This is easier to follow. And it's clearly visible what syntaxkind is related to what node type.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CloudNative/AzureFunctionsStateless.cs [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    {
+        StaticClass.Property = 42;              // Compliant
+    }
+
+    [FunctionName("Sample")]
+    public static void Write()
+    {
+        var local = 0;
+
+        Property = 42;          // Noncompliant {{Do not modify a static state from Azure Function.}}
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CloudNative/AzureFunctionsStateless.cs:70]: I don't see why that would be a good idea in Azure Funcion.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CloudNative/AzureFunctionsStateless.cs [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+    [Another("Something")]
+    public static void WithAnotherAttribute()   // Compliant
+    {
+        Property = 42;
+    }
+
+    public static void NoAttribute()
+    {
+        StaticClass.Property = 42;              // Compliant
+    }
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CloudNative/AzureFunctionsStateless.cs:63]: We prefer FNs over FPs, so I'll keep it as it is. Detecting azure-function-class would become too expensive with the current implementation

**Review by martin-strecker-sonarsource — APPROVED** [review]

**Comment by martin-strecker-sonarsource** [comment]
LGTM

**Pull request merged**